### ISSUE TITLE
Security Updates

### ISSUE THREAD

**Issue by citibeth** [LLNL/spack#562]
Today, I learned that a number of versions of git have a potentially serious security flaw.  I immediately revised git/package.py to disable the flawed versions, and to make it build good versions by default.  @tgamblin seemed to agree that this was a good idea, and he merged the PR.  So far so good.

The question is... what next?  How will all the Spack users who have already built git and are now using it be informed that they need to rebuild git, and anything that depends_on('git')?  And supposing that features are added to Spack that choose packages that have already been built... would that serve to further propagate dependence on bad git versions?

I think we need some kind of infrastructure to deal with this issue in a systematic and thorough fashion.  Here is a "first-draft" idea.

Suppose we have a git blacklist that is part of the Spack git repo.  The blacklist would be either:
1. As one big file (eg, blacklist.yaml)
2. (Probably better): Within each package.py file, allowing versions to be blacklisted.  For example:

```
class Git(Package):
    """Git is a free and open source distributed version control
       system designed to handle everything from small to very large
       projects with speed and efficiency."""
    homepage = "http://git-scm.com"
    url      = "https://github.com/git/git/tarball/v2.7.1"

    version('2.8.0-rc2', 'c2cf9f2cc70e35f2fafbaf9258f82e4c')
    version('2.7.3', 'fa1c008b56618c355a32ba4a678305f6')
    version('2.7.1', 'bf0706b433a8dedd27a63a72f9a66060')

    # See here for info on vulnerable Git versions:
    # http://www.theregister.co.uk/2016/03/16/git_server_client_patch_now/
    # All the following are vulnerable
    version('2.6.3', 'b711be7628a4a2c25f38d859ee81b423', blacklist=True)
    version('2.6.2', 'da293290da69f45a86a311ad3cd43dc8', blacklist=True)
    version('2.6.1', '4c62ee9c5991fe93d99cf2a6b68397fd', blacklist=True)
    version('2.6.0', 'eb76a07148d94802a1745d759716a57e', blacklist=True)
    version('2.5.4', '3eca2390cf1fa698b48e2a233563a76b', blacklist=True)
    version('2.2.1', 'ff41fdb094eed1ec430aed8ee9b9849c', blacklist=True)
```

Spack would then have a command ("spack update"?) that would identify any currently-built packages that are now blacklisted, and do what it takes to get rid of them.  This would involve re-building anything that depends on them.   It is not immediately clear (to me) where or not this step would require manual intervention.  Suppose we have installed A@1.5, and B@2.0 where B depends_on('A@1.5').  And suppose A@1.5 is blacklisted.  It now becomes impossible to rebuild B when we get rid of the blacklisted A@1.5.  Hopefully, B/package.py will have been changed to depends_on some other version of A.

If we want people to notice and respond to blacklisted versions, then it will be important to upgrade our package definitions regularly.  Some issues this brings up:
1. Currently, that means upgrading all of Spack with "git update."  Maybe that's not so bad.  The alternative is to update package definitions separately from spack.
2. Probably best to regularly update released Spack versions with security fixes in the package files.  This would need to happen on some regular schedule.

**Comment by tgamblin** [comment]
I like the idea of adding a `blacklist` attribute and of notifying the user of packages that are blacklisted.  Incidentally, we need the same feature for a regular old update -- i.e., "tell me if there are newer stable versions of things (and rebuild them)" or "tell me if there are bleeding-edge versions (and rebuild them)".

Spack needs a better notion of a current stack -- i.e. a set of versions that are "recommended" by the community and that get updated frequently like this.  I was actually going to put this on the agenda for the telcon today -- @mplegendre and I had a discussion yesterday about how `packages.yaml` preferences should work w.r.t. spack default, site-specific, and user-specific preferences, and it's related.  Having a blacklist like this would be a good addition to the scheme we were discussing.

Anyway thanks for the suggestion -- if you're on the telcon we can talk more about it then, or the proposal will be in the notes.  I can add a blacklist in to the requirements.

**Comment by citibeth** [comment]
OK, how about the following tags in the version() method:

   *) blacklist: Spack will never install this automatically, and will protest if you ask for it specifically.  Getting it installed will require an override switch.
   *) deprecated: Spack will only protest mildly if you try to install this, but will still go ahead and install.
   *) auto_upgrade_to=new_version: If Spack wants to upgrade, it will feel confident to auto-upgrade to this version.  (eg: version('7.0.0', 'abcd123', auto_upgrade_to='7.0.5')).
   *) manual_upgrade_to=new_version: This is a recommended upgrade, but might introduce some incompatibility.  (eg: GCC  version('4.9.3', 'abcd123', manual_upgrade_to='5.1')

Spack would then have the following upgrade commands:

 *) "spack upgrade blacklisted": Upgrade ONLY blacklisted packages, to the version specified by auto_upgrade_to
  *) "spack upgrade deprecated": Also upgrade deprecated packages, to the auto_upgrade version
  *) "spack upgrade all": Also upgrade anything with an auto_upgrade version listed

All of these would have an optional "manual" version, which tries to upgrade to the manual_upgrade version instead.

Problems:

1. Do we need to fix up packages.yaml when we do this?

**Comment by citibeth** [comment]
After thinking about it and our conference call March 17, I suggest the following set of discrete features:

1. Assume that packages might become broken because their (previously installed) dependencies are removed, and beef up support for that.  For example... suppose that netcdf depends on hdf5.  If I do "spack uinstall -f hdf5," this will put netcdf into an unusable state.  Spack should be able to find all unusable packages and mark them as such.  (They should appear as unusable in "spack find."  "Spack load" should not load them.  ETc).

2. There should be a nice top-level script that, given a bunch of specs, does a "spack uninstall -f" on all of them.  Sysadmins regularly get blacklists from some or another advisory.  This program will be run against those blacklists to ensure a Spack installation is in compliance.

3. Spack should be able to remember blacklists given to it (in (2) for example), and prevent repeat installation of those packages/versions in the future.

4. One should also be able to specify blacklisted versions in packages.py.  Thus, Spack will learn about blacklisting from sysadmin blacklists, or from other Spack users.  Either way, the response to a newly blacklisted package should be to immediately uninstall it.

5. As Todd mentioned, we should be able to (essentially) symlink one package into another.  For example, link netcdf-4.1.2 --> netcdf-4.1.1.  But I'd like to see a bit more rigor here; for example, what happens when someone uninstalls netcdf-4.1.1?
    I would therefore suggest we have PHYSICAL and LOGICAL packages.  Users use LOGICAL packages, and they are backed by PHYSICAL packages.  Spack as we know it today installs a logical and physical package together.  But now we can separate them as needed .  For example:
     (logical) netcdf-4.1.2 ---> (physical) netcdf-4.1.1
     (logical) netcdf-4.1.1 ---> (physical) netcdf-4.1.1
Spack would keep refernece counts of logical packages referencing physical packages.  When you install a logical package, it decreases the physical package's reference count; and uninstalls the physical package only if the ref count has gone to zero.  There need to be Spack commands to explicitly manipulate and maintain the linkage relationship between logical and physical libraries.

6. Linux shared libraries indicate ABI compatibility through version numbering conventions.  For example.... we have libc.so.5, libc.so.5.1, libc.so.5.1.2.   I believe that changes in the third part of the version number retain ABI compatibility.  Spack cannot rely on this convention; but we CAN add information to package.py files explaining which versions are API or ABI compatible with each other.  This will help Spack make automatic repair decisions.

7. Once blacklisted packages have been uninstalled, that will leave a lot of holes in the pyramid.  Those holes are to be filled through a Spack repair process.  For each hole, Spack would try the following options:
   a) Look for a version of the package (installed or not yet installed) that is ABI-compatible with the hole.  If that can be found, build  & install that package (if not already installed), and then link the (logical) hole to the (physical) package just installed.  Then look for disabled packages, and re-enable any that are now complete (in their dependencies).
   b) If no ABI-compatible package version is available, look for an API-compatible version instead.  Build and install the API-compatible version, and link the (logical) hole to the (physical) package just installed.  Now uninstall all packages that depend on the library just replaced (since they now must be rebuilt).  They will be rebuilt through repeated application of this repair process.
   c) If no ABI-compatible or API-compatible package version is available, the hole cannot be filled automatically.  The user will have to find a new version, then re-install anything that depends on it.

**Comment by dshrader** [comment]
I am afraid I haven't been able to call in to the last few telecons, so I have missed those conversations. So, I am pretty much coming at this just with the information here in this post.

What kind of "blacklists" are we talking about here? The ambiguity makes me extremely nervous when talking about a spack installation that is in place for all users of a system, not just a single user. In that situation, it is not acceptable to automatically change out a package that users link against for releases of their code, even if the users get a "replacement." Depending on the code, there may be governmental certifications or other such oversight/bureaucracy that is part of an agreement between the users of the system and those who provide software on the system. Unless we can guarantee that all codes everywhere can always get the same answer when we "replace" a package, I don't think it is a good idea to promote automatic removal and/or replacement.

What is the rationale for using sym links such as what was referenced above: netcdf-4.1.2 ---> netcdf-4.1.1? Under what circumstances is it ok to tell users that they are getting 4.1.2 but they are actually getting 4.1.1? I especially don't like this if the version that is now linking to another version was actually installed at some point.

API/ABI compatibility isn't always there even when the vendors state that it is. For example, Fortran code produced by intel 16.0.0.026 won't work with later intel 16 updates under certain circumstances, even though intel says they are all ABI compatible. Unless we are going to test for API/ABI compatibility ourselves, I don't know how much of this game we want to play. Perhaps we have to be more conservative when it comes to Fortran and C++ since they don't have external standards for the API like C does.

As I reread my post, I guess I am mostly concerned about the nature of the "blacklist." I think everything I rambled on about could be taken care of by more fully understanding the "blacklist." Where does it come from? What packages end up on there? Git seems to be one of them, but do compilers show up on the blacklist? If there is even a small chance that a package as basic as a  compiler can end up on the "blacklist," I feel extreme caution is in order.

**Comment by alalazo** [comment]
@citibeth That seems to me overly complicated. Not to mention that doing the damage in advance (`uninstall -f`) and try to repair it later may lead to pain if some of the packages couldn't be built with the replacement version. 

I also don't like having a LOGICAL `netcdf 4.1.2` that is a PHYSICAL `netcdf 4.1.1`. At first glance it seems it could add a layer of confusion. What should a user think when prompting : 
```
spack install netcdf@4.1.2
```
he sees that `4.1.1` gets downloaded?

**Comment by citibeth** [comment]
David,

Thank you for your comments.  AFAIK, this issue has not been discussed much
at telcons, so you haven't missed anything.

As I reread my post, I guess I am mostly concerned about the nature of the
> "blacklist." I think everything I rambled on about could be taken care of
> by more fully understanding the "blacklist." Where does it come from? What
> packages end up on there? Git seems to be one of them, but do compilers
> show up on the blacklist? If there is even a small chance that a package as
> basic as a compiler can end up on the "blacklist," I feel extreme caution
> is in order.
>

The idea is that when a security vulnerability is discovered in a
particular version of a particular piece of software, then that information
gets transmitted to a Spack installation in an efficient manner.  I admit
that security is not such an issue if you're running mantle convection
simulations on a supercomputer behind a firewall.  But I think that Spack
also has applications in Internet-facing production systems.

For example... in our installation, we are running a super-outded server
daemon, that also now has known vulnerabilities.  Our sysadmin is reluctant
to deploy a newer/more secure version of that server  because he would have
to build it himself.  Unfortunately, no RPMs of a modern version of this
server have yet surfaced.  I think we would have more flexibility to deploy
the versions of software that would suit our needs, if we were running
Spack.  And once I started down that line of thinking, the first issue was,
how do we manage that Spack installation in the face of a constant barrage
of security alerts?

I could think of two ways that vulnerability information would get to our
hypothetical production Spack installation:

   1. The government periodically sends us a list of vulnerable software,
and it is our sysadmins' task to ensure that none of the vulnerable
versions are running on our systems.  The attitude is "we don't care what
we break, we just have to remove the bad stuff."  The idea of doing a
"spack uninstall -f" on all vulnerable versions would comply with this
need, leaving the (harder) problem of fixing the stuff you just broke for
later (that's the repair process I outlined).

   2. Vulnerability information can also come from Spack itself.  In the
case of git... when I read that many versions are vulnerable, I updated the
Spack recipe based on that information.  However, this does nothing for all
the Spack users who already have vulnerable versions of git installed.
There needs to be some way to tell Spack, not just to stop building the
vulnerable version, but also to consider replacing existing vulnerable
installations.  Hence the idea of blacklisting inside the package.py files.

I think it's clear from this discussion that:
  a) Blacklist information needs to be treated differently depending on
whether it came from Spack, or from some blacklist file supplied by the
user.
  b) The user needs to maintain control over what happens to blacklisted
software.

If there is even a small chance that a package as basic as a compiler can
> end up on the "blacklist," I feel extreme caution is in order.


The reality is, anything can end up on a blacklist.  Suppose that hackers
affiliated with some foreign power infiltrate GCC or LLVM.  Suppose they
manage to get code in those compilers that turns anything compiled with
them into a virus.  In that case, yes, the compilers would go on the
blacklist.  Actually, you'd probably have to rebuild your whole system.
This has been demonstrated, and it's kind of scary:

http://c2.com/cgi/wiki?TheKenThompsonHack

Now on to the issue of repairing binaries after a blacklisted package has
been removed...

> In that situation, it is not acceptable to automatically change out a
> package that users link against for releases of their code, even if the
> users get a "replacement."... Unless we can guarantee that all codes
> everywhere can always get the same answer when we "replace" a package, I
> don't think it is a good idea to promote automatic removal and/or
> replacement.
>
Spack cannot make such guarantees.  What it CAN do is act as a repository
for our best knowledge of which versions are ABI-compatible and
API-compatible with which other versions.  That information would allow
Spack to auto-replace versions, if the user so chooses.  The paranoid Spack
user will never trust ABI compatibility and always rebuild with new
versions.  That should be one option.  But right now, Spack doesn't provide
that kind of information at all, so there's not much we can do.

By formalizing the idea of logical vs. physical packages (rather than just
hacking some symlinks), Spack can be up-front about where in your DAG
replacements have happened.  "spack spec" would do this.  You can then
decide what to do about it.  Spack should never "silently" replace one
package with another, you should aways be told what is going on.

Depending on the code, there may be governmental certifications or other
> such oversight/bureaucracy that is part of an agreement between the users
> of the system and those who provide software on the system.
>
If the blacklist came from the government, then you will not be allowed to
continue running it while maintaining government certification.


> What is the rationale for using sym links such as what was referenced
> above: netcdf-4.1.2 ---> netcdf-4.1.1? Under what circumstances is it ok to
> tell users that they are getting 4.1.2 but they are actually getting 4.1.1?
> I especially don't like this if the version that is now linking to another
> version was actually installed at some point.
>
Symlinks are an implementation tactic.  The idea is to have logical vs.
physical packages.  You depnd on / link to a logical package, which is
actually a (possibly different) physical package under the hood.  Another
way to do this without symlinks would be to simply build 4.1.1 and put it
in a directory called 4.1.2.  Either way, the idea is that Spack will tell
you what's going on, and where these changes have taken place.

API/ABI compatibility isn't always there even when the vendors state that
> it is. For example, Fortran code produced by intel 16.0.0.026 won't work
> with later intel 16 updates under certain circumstances, even though intel
> says they are all ABI compatible.
>
We are mixing up ABI compatibility of object code produced by different
compiler versions, vs. ABI compatibility of different libraries.  The
concern here is about ABI compatibility of libraries --- for example, is
version 1.2.1 of OpenSSL ABI-compatible with version 1.2.2?  Spack always
assumes incompatibility when compiler versions are changed.

Different projects go to different amounts of effort to maintain ABI
compatibility.

http://openssl.6102.n7.nabble.com/Verification-of-ABI-compatibility-td62705.html

Projects like OpenSSL are very careful; projects like IceBin (mine) not at
all.  We would have to use our discretion when putting ABI and API
compatibility information into Spack.

Unless we are going to test for API/ABI compatibility ourselves,
>
Maybe we should:

https://lvc.github.io/abi-compliance-checker/

-- Elizabeth

**Comment by adamjstewart** [comment]
This might be a dumb question, but is there currently any way to do the following?:

I install hdf5@1.8.16 and a bunch of packages that depend on it. I later realize that I installed hdf5@1.8.16 incorrectly, or there is a security flaw, or someone adds an extra variant and the hash changes. I want to uninstall hdf5 **and all of the packages that depend on it**, not just force the uninstall. It would also be cool to be able to rebuild hdf5 and all of the packages that depended on it (either the same version or a new version without security flaws).

If we can add the ability to do this, `spack uninstall -f` won't be necessary anymore, right?

**Comment by citibeth** [comment]
After talking with our sysadmins, I think the biggest problem is getting accurate blacklist information.  Typically, packages are blacklisted by OS vendors: they have a team of people who keep their ears open and constantly produce OS patches.  Those patches then get pushed to users of the OS.

The same could happen with Spack, of course.  But it would require that someone watch security announcements and translate them into whatever blacklist / updates Spack might decide upon.  Even piggy-backing off, say RHEL or Debian security update announcements, would require thought and diligence.  Right now, no one has that job, so there's no way to expect it would get done in a consistent manner.

Unless/until there's a reliable way to obtain Spack blacklist information, I don't think Spack would be suitable for providing the software that runs network-facing servers.

**Comment by dshrader** [comment]
I am a proponent of the user making the choice on what to do with blacklisted packages. In general, though, given the amount of variability in this space, it might be too big of a thing for Spack to tackle, as you mentioned @citibeth . A vulnerability in git isn't the same thing as a vulnerability in openssl (unless, of course, it is!).

Perhaps it might be better to punt this one to the users of Spack. They had to keep up with acceptable versions of their software and its dependencies before Spack, we may just have to have them keep doing it with Spack. Of course, we can make it easier for them to build replacements, so it may still be beneficial to provide means by which software can be removed and/or replaced at the user's request.

**Comment by citibeth** [comment]
>
> Perhaps it might be better to punt this one to the users of Spack. They
> had to keep up with acceptable versions of their software and its
> dependencies beforeSpack, we may just have to have them keep doing it
> with Spack. Of course, we can make it easier for them to build
> replacements, so it may still be beneficial to provide means by which
> software can be removed and/or replaced at the user's request.


I agree, that's probably the only practical approach at this point.  What
it means is that Spack will be useful for HPC (its main purpose), but not
for maintaining Internet-facing servers.

**Comment by tgamblin** [comment]
Catching up on the conversation.

> 1. Assume that packages might become broken because their (previously installed) dependencies are removed [...] Spack should be able to find all unusable packages and mark them as such.  (They should appear as unusable in "spack find."

This is already tracked.  `spack find --missing` will show both installed packages *and* packages needed by something installed.  The database in `opt/spack/.spack-db` keeps a reference count on installation records, and a record is only removed when nothing depends on it.  `installed` is a separate field in the DB -- if you `uninstall -f` something that has dependents, it is marked not installed but its ref count will be positive.

> 2. There should be a nice top-level script that, given a bunch of specs, does a "spack uninstall -f" on all of them.  Sysadmins regularly get blacklists from some or another advisory.  This program will be run against those blacklists to ensure a Spack installation is in compliance.

Maybe.  Personally I think stuff should get upgraded at the same time.  i.e. there should just be an "update" command, and updating blacklisted things should be an option.

> 3. Spack should be able to remember blacklists given to it (in (2) for example), and prevent repeat installation of those packages/versions in the future.

I'm not sure this is needed in addition to blacklists in `package.py` and `packages.yaml`.

> 4. Either way, the response to a newly blacklisted package should be to immediately uninstall it.

This is up to users.  Spack can certainly warn, but Spack shouldn't just uninstall things without asking.  It's up to the user to evaluate the risk/reward.

> 5. As Todd mentioned, we should be able to (essentially) symlink one package into another.

I would incorporate this into a general update command, and I would integrate it with the uninstalling above.

>     I would therefore suggest we have PHYSICAL and LOGICAL packages.  Users use LOGICAL packages, and they are backed by PHYSICAL packages.  

We have talked about things like this in the discussion of views in #568 and when we've talked about having environments (like nix/guix profiles or conda environments). I think there are lots of examples out there of how to manage this in a package manager.

> 6. Linux shared libraries indicate ABI compatibility through version numbering conventions.  For example.... we have libc.so.5, libc.so.5.1, libc.so.5.1.2.   I believe that changes in the third part of the version number retain ABI compatibility.  Spack cannot rely on this convention; but we CAN add information to package.py files explaining which versions are API or ABI compatible with each other.  This will help Spack make automatic repair decisions.

We can do this type of analysis with RedHat's `libabigail` and `abidiff`.  I've talked to some RedHat people about using `abidiff` to pick a version to "upgrade" to.  Hints in the package could be nice, too.

> 7. Once blacklisted packages have been uninstalled, that will leave a lot of holes in the pyramid.  Those holes are to be filled through a Spack repair process.  For each hole, Spack would try the following options:
>    a) Look for a version of the package (installed or not yet installed) that is ABI-compatible with the hole.  If that can be found, build  & install that package (if not already installed), and then link the (logical) hole to the (physical) package just installed.  Then look for disabled packages, and re-enable any that are now complete (in their dependencies).
>    b) If no ABI-compatible package version is available, look for an API-compatible version instead.  Build and install the API-compatible version, and link the (logical) hole to the (physical) package just installed.  Now uninstall all packages that depend on the library just replaced (since they now must be rebuilt).  They will be rebuilt through repeated application of this repair process.
>    c) If no ABI-compatible or API-compatible package version is available, the hole cannot be filled automatically.  The user will have to find a new version, then re-install anything that depends on it.

Note that all of the above is very similar to what is necessary for concretizing using already-installed dependencies.  We are in the process of developing more general mechanisms for this stuff.  It all reduces to constraint solving, and I think once that is in place this type of thing becomes a lot easier.

**Comment by tgamblin** [comment]
@adamjstewart: 

> This might be a dumb question, but is there currently any way to do the following?:
> 
> I install hdf5@1.8.16 and a bunch of packages that depend on it. I later realize that I installed hdf5@1.8.16 incorrectly, or there is a security flaw, or someone adds an extra variant and the hash changes. I want to uninstall hdf5 **and all of the packages that depend on it**, not just force the uninstall. It would also be cool to be able to rebuild hdf5 and all of the packages that depended on it (either the same version or a new version without security flaws).
> 
> If we can add the ability to do this, `spack uninstall -f` won't be necessary anymore, right?

Yes, there are basically two ways you could handle upgrades in Spack:
  1. Rebuild everything, which can actually be done *alongside* what is already installed.  @alalazo recently sent his PR #457  recursive uninstall, which would make this easier.  
  2. Use symlinks (or some other essentially in-place upgrade mechanism) to swap packages out from under existing installations.

Both seem good to have.  (1) is a lot more expensive, which makes (2) appealing.  In either case you really want to have the ability to roll back so that your upgrade can be tested.  Both of these mechanisms have been done in package managers like Nix and Guix -- they use a "profile" that is tracked over time.  The profile looks a lot like a single merged prefix, and it functions as a set of ref-count roots for the raw installations.  Spack really only has the raw installs right now, along with module generation.  Adding something like a profile is on the list of things I really want, and I would do it so that it has something resembling transactional rollback.  I don't really think we need to hash out the technical details here.

**Comment by tgamblin** [comment]
> > Perhaps it might be better to punt this one to the users of Spack. They
> > had to keep up with acceptable versions of their software and its
> > dependencies beforeSpack, we may just have to have them keep doing it
> > with Spack. Of course, we can make it easier for them to build
> > replacements, so it may still be beneficial to provide means by which
> > software can be removed and/or replaced at the user's request.
> 
> 
> I agree, that's probably the only practical approach at this point.  What
> it means is that Spack will be useful for HPC (its main purpose), but not
> for maintaining Internet-facing servers.

FWIW, I think this is a good discussion, and @citibeth has a lot of very similar ideas about *how* this might be dealt with.  I think @dshrader's point about leaving the ultimate decision up to the user is important.  We cannot dictate to users/sites what their security *policies* should be.

However, we can certainly provide mechanisms that enable this stuff.  There are a few features that need to get integrated before we can have more seamless upgrade-like commands:
  1. Profiles/environments/views -- these are all ways to expose packages to users so that a "physical" package can be upgraded behind a logical view.
  2. Better constraint solving / searching what is installed for a satisfactory spec -- needed to facilitate deciding *which* packages to upgrade to.
  3. symlink grafts (or auto-rebuilding) -- needed to do in-place updates of Spack packages.

All of these are planned in one form or another.  I think we should revisit this discussion once they are in, as then we will have a very clear set of things we can do to enable upgrades, blacklists, etc.

**Comment by ax3l** [comment]
I think the underlying organizational issue is, that it is as cumbersome to add new "trusted" versions from the downstream point of view as it is to "blacklist" existing ones.

Did you think about adding a remote URL to a list of "trusted" versions and just adding that this file needs to be signed, e.g., with a once-added downstream maintainer gpg key? The downstream maintainers will anyway be quicker publishing what versions need to be fixed than we can ever be, following all packages at the same time without automation.

Basically everyone publishing on FTP servers is putting a signature alongside it anyway (see, e.g. your initial [git repo](https://www.kernel.org/pub/software/scm/git/) issue).
And git tag based releases can (should) be signed, too.

**Comment by citibeth** [comment]
We need to be careful about what we're guaranteeing here.  I think Spack should guarantee that the version downloaded by the user will be the same as the version downloaded by the Spack package author, as long as Spack itself hasn't been tampered with.

Relying on upstream publishers to provide signatures does not give this guarantee.  The Spack package author should use the upstream signatures to increase the assurance that the tarball the Spack package author sees is the same as the tarball intended by the upstream author.  However... since so many upstream signatures appear on the same website / FTP server beside the tarball itself, they provide little or no actual security.  (I suppose they do help ensure something didn't go wrong in the download).

**Comment by alalazo** [comment]
This issue relates to #172 . I think we should close one of the two and reference it in the other.

**Comment by citibeth** [comment]
My conclusion from the whole discussion is this isn't going to happen.  We just don't have the people or expertise to be chasing security flaws like a "real" Linux distro.  (Updating in place won't happen either.  That's a hard problem, and rebuilding with Spack really isn't so bad).

I think our best bet here is external repo adapters (#2777).  That would allow us to piggy-back security updates off of a larger project with dedicated maintenance staff.

The `buildable: False` stuff also helps.  We can rely on system libraries for (say) OpenSSL; again, foisting the security burden on somebody else.

**Comment by citibeth** [comment]
I say close #172 also.
